L1Trigger/L1THGCal/src/be_algorithms/HGCalTowerMap2DImpl.cc

```diff
@@ -0,0 +1,155 @@
+///
+/// \class HGCalTowerMap2DImpl
+///
+/// \author: Thomas Strebler
+///
+/// Description: first iteration of HGCal Tower Maps
+
+#include "L1Trigger/L1THGCal/interface/be_algorithms/HGCalTowerMap2DImpl.h"
+#include "DataFormats/Math/interface/normalizedPhi.h"
+
+constexpr double HGCalTowerMap2DImpl::kEtaMax_;
+constexpr double HGCalTowerMap2DImpl::kEtaMin_;
+constexpr double HGCalTowerMap2DImpl::kPhiMax_;
+constexpr double HGCalTowerMap2DImpl::kPhiMin_;
+
+
+HGCalTowerMap2DImpl::HGCalTowerMap2DImpl( const edm::ParameterSet& conf ) :
+  nEtaBins_(conf.getParameter<unsigned>("nEtaBins")),
+  nPhiBins_(conf.getParameter<unsigned>("nPhiBins")),
+  etaBins_(conf.getParameter<std::vector<double> >("etaBins")),
+  phiBins_(conf.getParameter<std::vector<double> >("phiBins")),
+  layerWeights_(conf.getParameter< std::vector<double> >("layerWeights"))
+{
+  edm::LogInfo("HGCalTowerMap2DImpl") << "Number of eta bins for the tower maps: " << nEtaBins_<<endl;
+  edm::LogInfo("HGCalTowerMap2DImpl") << "Number of phi bins for the tower maps: " << nPhiBins_<<endl;
+
+  if(etaBins_.size()>0 && etaBins_.size()!=nEtaBins_+1){
+    edm::LogError("HGCalTowerMap2DImpl") << "nEtaBins for the tower mapsnot consistent with etaBins size"<<endl;
+  }
+  if(phiBins_.size()>0 && phiBins_.size()!=nPhiBins_+1){
+    edm::LogError("HGCalTowerMap2DImpl") << "nPhiBins for the tower mapsnot consistent with phiBins size"<<endl;
+  }
+    
+  //If no custom binning specified, assume uniform one
+
+  if(etaBins_.size()==0){
+    double eta_step = (kEtaMax_ - kEtaMin_)/nEtaBins_;
+    for(unsigned int i=0; i<nEtaBins_; i++)
+      etaBins_.emplace_back( kEtaMin_ + eta_step*i );
+    etaBins_.emplace_back( kEtaMax_ );
+  }
+
+  if(phiBins_.size()==0){
+    double phi_step = (kPhiMax_ - kPhiMin_)/nPhiBins_;
+    for(unsigned int i=0; i<nPhiBins_; i++)
+      phiBins_.emplace_back( kPhiMin_ + phi_step*i );
+    phiBins_.emplace_back( kPhiMax_ );
+  }
+
+  edm::LogInfo("HGCalTowerMap2DImpl") << "Eta bins for the tower maps: {";
+  for(auto eta : etaBins_) edm::LogInfo("HGCalTowerMap2DImpl") << eta << ",";
+  edm::LogInfo("HGCalTowerMap2DImpl") << "}" <<endl;
+  edm::LogInfo("HGCalTowerMap2DImpl") << "Phi bins for the tower maps: {";
+  for(auto phi : phiBins_) edm::LogInfo("HGCalTowerMap2DImpl") << phi << ",";
+  edm::LogInfo("HGCalTowerMap2DImpl") << "}" <<endl;  
+
+}
+
+
+
+void HGCalTowerMap2DImpl::buildTowerMapClusters(const std::vector<edm::Ptr<l1t::HGCalTriggerCell>> & triggerCellsPtrs,
+					     l1t::HGCalClusterBxCollection & clusters
+				       ){
+
+  map<pair<pair<int,int>,int>, l1t::HGCalCluster> ClusterTowerMap; //key = pair(iEtaPhiBin,layer)	
```

Comment:
It would be better to encapsulate the tower map into a class with a more user-friendly interface. You would have here one tower map per layer (a vector of tower map objects).
To simplify the merging of tower maps in multiple layers, you could implement a `+` operator.